### PR TITLE
Refactor http library to provide more granular helpers

### LINKED ISSUE #939: Http library more granular helpers to get route and http infromation of operation
It would be great if you can pass a generic `namespace`, `interface` or `operation` and be able to get the same info returned by `getAllRoutes`

### PR DESCRIPTION
**Pull request by timotheeguerin**
resolves #939 

## Http library function changes

Additions
```ts

/**
 * Return the Http Operation details for a given Cadl operation.
 * @param operation Operation
 * @param options Optional option on how to resolve the http details.
 */
export function getHttpOperation(
  program: Program,
  operation: Operation,
  options?: RouteOptions
): [HttpOperation, readonly Diagnostic[]];

export function getHttpOperations(
  program: Program,
  container: OperationContainer,
  options?: RouteOptions
): [HttpOperation[], readonly Diagnostic[]];

/**
 * Returns all the services defined.
 */
export function getAllHttpServices(
  program: Program,
  options?: RouteOptions
): [HttpService[], readonly Diagnostic[]];
```

Deprecation:
```ts
/**
 * @deprecated use `getAllHttpServices` or `resolveHttpOperations` manually
 */
export function getAllRoutes(
  program: Program,
  options?: RouteOptions
): [HttpOperation[], readonly Diagnostic[]];
```

### CODE REVIEW

**Comment by daviwil** [comment]
Is it possible this PR addresses the requests in https://github.com/microsoft/cadl/issues/685?  It seems that you're gathering path fragments upward instead of pushing them downward like I was before.  This might make routes more composable.

File: packages/compiler/core/helpers/operation-utils.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+   */
+  recursive?: boolean;
+}
+
+/**
+ * List operations in the given container. Will list operation recursively by default(Check subnamespaces.)
+ * @param container Container.
+ * @param options Options.
+ */
+export function listOperations(
```
[the remark this answers is not in the document]
Comment (timotheeguerin) [reply packages/compiler/core/helpers/operation-utils.ts:17]: that's fair, `findOperations` looks good to me. think we are discussing this tomorrow during the design meeting, so will just see if everyone is onboard with all the names

**Comment by timotheeguerin** [comment]
> Is it possible this PR addresses the requests in #685? It seems that you're gathering path fragments upward instead of pushing them downward like I was before. This might make routes more composable.

I don't think so, if we go with the `@routeOverride` way then this doesn't change anything (i think) and if we go with the interface composition way, it doesn't change much right now. Main problem being the decorator of interface gets lost when using extends. But maybe we could, and I think we should have a `sourceOperation` property on operation that show what was the original operation so you can trace back. With that we could potentially get the original interface back and get the route defined for it.

Another approach I have been thinking is maybe what we want is also a `include` interface which just copy the interface under the namespace or other interface. In this PR https://github.com/Azure/cadl-azure/pull/2005 you end up having to do `interface Operations extends Azure.ResourceManager.Operations {}` just to include the other interface. I think if we could just have `include  Azure.ResourceManager.Operations;` that would be nice.

**Comment by daviwil** [comment]
I like that `include` syntax, would certainly be useful for pulling in standard patterns that don't need a unique name.

**Comment by daviwil** [comment]
Looks like this is good to merge!

File: packages/compiler/core/helpers/operation-utils.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (timotheeguerin) [reply packages/compiler/core/helpers/operation-utils.ts:17]: what about `listOperationsIn`?

**Pull request merged**